tests/TestSourceMaps.hs

```diff
@@ -4,45 +4,65 @@ import Prelude
 
 import Control.Monad (forM_)
 import Test.Hspec
-import System.FilePath (replaceExtension, (</>), takeFileName)
+import System.FilePath (replaceExtension, (</>), takeFileName, dropExtensions)
 import qualified Language.PureScript as P
 import qualified Data.ByteString as BS
 import TestUtils (goldenVsString, modulesDir, getTestFiles, SupportModules (..), compile')
 import qualified Data.Set as Set
 import TestCompiler (getTestMain)
+import System.Directory (doesFileExist)
 
 spec :: SpecWith SupportModules
 spec =
   goldenFiles
 
+-- Test module names must be prefixed with "SourceMaps." and reside under tests/purs/sourcemaps. 
+--  eg. Module file tests/purs/sourcemaps/Test123.purs must be named SourceMaps.Test123  
+-- Modules created for the purpose of testing specific github issues should be named SourceMaps.Bug<issue_number>. 
+--  eg. File name tests/purs/sourcemaps/Bug123.purs, module name SourceMaps.Bug123
+--
+-- Test modules and compilation output can be conveniently copied to .source-maps/<module_name>/
+--  by running the get-source-maps.sh script.
 goldenFiles :: SpecWith SupportModules
 goldenFiles = do
   sourceMapsFiles <- runIO $ getTestFiles "sourcemaps"
 
   describe "golden files" $
-    forM_ sourceMapsFiles $ \testPurs ->
-      it ("'" <> takeFileName (getTestMain testPurs) <> "' should compile to expected output") $ \support ->
-        assertCompilesToExpectedOutput support testPurs
-
-compilationOptions :: P.Options
-compilationOptions = P.defaultOptions { P.optionsCodegenTargets = Set.fromList [P.JS, P.JSSourceMap] }
+    forM_ sourceMapsFiles $ \inputFiles ->
+      it ("'" <> takeFileName (getTestMain inputFiles) <> "' should compile to expected output") $ \support ->
+        assertCompilesToExpectedOutput support inputFiles
 
 assertCompilesToExpectedOutput
   :: SupportModules
   -> [FilePath]
   -> Expectation
 assertCompilesToExpectedOutput support inputFiles = do
+
+  let
+    modulePath = getTestMain inputFiles
+    moduleName = "SourceMaps." <> (dropExtensions . takeFileName $ modulePath)
+    compiledModulePath = modulesDir </> moduleName </> "index.js.map"
+
   (result, _) <- compile' compilationOptions False support inputFiles
```

Comment:
I'm thinking this `Boolean` flag here that does the `Main` module name check could instead be `Maybe ExpectedModuleName`, and we'd check the module name satisfies that check. So, usages of `False` become `Nothing`. Usages of `True` become `Just IsMain` and this usage becomes `Just IsSourceMap`, which verifies the module starts with `SourceMaps.` and the latter part matches the file name.

Reply:
I'm looking into this idea and I think it makes a lot of sense. But since the caller also needs to know the full path information of the compiled file, I'm thinking the `compile'` function should also return it so that it doesn't need to be calculated in both places. 
What's your opinion?

Reply:
Yeah, that sounds good.

Reply:
One thing I would like to do is compare the current source maps outputted by `master` and what this PR does. So, could you push changes that updates `compile'` to follow the design proposed here? Then we can compare them. I'd like to use a more complex example than what's currently being tested just to see how this change plays with everything else.

Reply:
FYI I've just pushed my latest changes. 
Now that you mentioned "more complex examples", I'm wondering if I should also include tests that exercise language features different than the ones that will be affected by this PR...

Reply:
> I'm wondering if I should also include tests that exercise language features different than the ones that will be affected by this PR...

That's partly what I had in mind. Do the changes here leak into other source maps? Or does it only affect the cases you're trying to address in this PR?

Reply:
How about a test case like this for seeing whether things "leak"?
```purs
module SourceMaps.Leakage where

import Prelude
import Data.Maybe (Maybe(..), fromMaybe)
import Data.Tuple (Tuple(..))
import Type.Proxy (Proxy(..))

import Data.Int as Int

data Dtor = Dtor Int Int
newtype Ntor = Ntor (Tuple Int Int)
type IntType = Int


foo :: String -> String -> String -> String -> String
foo s1 s2 sUnused s4 = fromMaybe "" do
  i <- Int.fromString s1
  whole@{ p } <- pure { p: 1, h: 2 }
  let
    mutuallyRecursive1 k b
      | k <= 0 = b
      | otherwise = mutuallyRecursive2 (k - 1) b
      
    mutuallyRecursive2 k b
      | k <= 0 = b
      | otherwise = mutuallyRecursive1 (k - 1) b
      
    -- s4 here shadows s4 arg above
    patternGuards s4
      | s4 == "a string" = s4
      
      | s4 == "another string"
      , s4 /= "" = s4
      
      | Just anotherInt <- Int.fromString s4
      , anotherInt `mod` 2 == 0 = s4
      
      | otherwise = s4
      
    caseOfInt = case _ of
      1 -> 1
      _ -> 2
      
    caseOfNum = case _ of
      1.0 -> 1
      _ -> 2
      
    caseOfNegInt = case _ of
      (-1) -> 1
      _ -> 2
      
    caseOfNegNum = case _ of
      (-1.0) -> 1
      _ -> 2
      
    caseOfString = case _ of
      "1" -> 1
      _ -> 2
      
    caseOfChar = case _ of
      '1' -> 1
      _ -> 2
      
    caseOfBool = case _ of
      true -> 1
      false -> 2
      
    caseOfArr = case _ of
      [] -> 1
      ["foo"] -> 2
      _ -> 3
      
    caseOfRec = case _ of
      { alpha, beta: 8 } -> alpha
      { beta: 1 } -> 2
      _ -> 3
      
    caseOfDtor = case _ of
      Dtor 1 _ -> 1
      Dtor x@1 2 -> x
      Dtor _ y -> 2 + y
      
    caseOfNtor = case _ of
      Ntor (Tuple 1 _) -> 1
      Ntor (Tuple x@1 2) -> x
      Ntor (Tuple _ y) -> 2 + y
      
  _ <- pure s4
  res <- bar 4 (pure [1, 2, 3])
  let
    extra = show $ map show
      [ caseOfInt 4
      , caseOfNegInt (-4)
      , caseOfNum 4.0
      , caseOfNegNum (-4.0)
      , caseOfString "afadf"
      , caseOfChar 'x'
      , caseOfBool true
      , caseOfArr []
      , caseOfRec { alpha: 0, beta: 4 }
      , caseOfDtor $ Dtor 8 10
      , caseOfNtor $ Ntor $ Tuple 1 8
      ]
      
    foo = "shadowed name"
    unusedBinding = "unused"
  pure $ show i <> patternGuards "abc" <> extra <> foo <> show res
      
  where
  bar :: forall c f a. Show a => Functor f => Monad c => IntType -> c (f a) -> c (f String)
  bar = baz
    where
    baz i c = do 
      f <- c
      pure $ map (\a -> show a <> " " <> show i) f

class MultiDimensionalDrift :: Type -> Symbol -> Row Type -> Int -> Constraint
class MultiDimensionalDrift ty sym row int | ty sym -> row int where
  example :: ty -> Proxy sym -> { | row } -> Proxy int

instance MultiDimensionalDrift Int "Int" row 0 where
  example _ _ _ = Proxy

Reply:
I think this is great! I added a couple extra things and I'm including this to the golden tests. 
But I'm thinking this could make new PRs (especially ones that optimize away code from index.js) to fail their tests. Not sure it that's a good or bad thing at this stage. 😄